### PR TITLE
Add spherical domains for CNSE3D model

### PR DESCRIPTION
**Pull request by blallen**
### Description

This PR adds everything necessary for running a hydrostatic balance test on the sphere with the CNSE3D model, as well as testing tracer conservation. Some interface changes were made to the FluidBC to make them GPU friendly and additional domains were added to handle the sphere. The core numerics change to get these tests running properly were added in #2135.

- [x] Code follows the [style guidelines](https://clima.github.io/ClimateMachine.jl/latest/DevDocs/CodeStyle/) OR N/A.
- [ ] Unit tests are included OR N/A.
- [x] Code is exercised in an integration test OR N/A.
- [ ] Documentation has been added/updated OR N/A.

### CODE REVIEW

**Comment by blallen** [comment]
bors try

**Comment by blallen** [comment]
bors try

File: test/Numerics/DGMethods/compressible_navier_stokes_equations/three_dimensional/ThreeDimensionalCompressibleNavierStokesEquations.jl [hunk 1]
```diff
@@ -279,8 +279,8 @@ end
     ρu = state.ρu
     ρθ = state.ρθ
 
-    u = ρu / ρ
-    θ = ρθ / ρ
+    u = ρu
+    θ = ρθ
```
Comment (blallen) [inline test/Numerics/DGMethods/compressible_navier_stokes_equations/three_dimensional/ThreeDimensionalCompressibleNavierStokesEquations.jl:283]: @sandreza why did you need to do this?

**Comment by blallen** [comment]
bors try

File: test/Numerics/DGMethods/compressible_navier_stokes_equations/three_dimensional/ThreeDimensionalCompressibleNavierStokesEquations.jl (hunk 1 above)
Comment (sandreza) [reply test/Numerics/DGMethods/compressible_navier_stokes_equations/three_dimensional/ThreeDimensionalCompressibleNavierStokesEquations.jl:283]: I was playing around with momentum diffusion, (which seems like the right thing to do in this case)

**Review by sandreza — APPROVED** [review]

File: test/Numerics/DGMethods/compressible_navier_stokes_equations/shared_source/grids.jl [hunk 2]
```diff
@@ -190,3 +190,116 @@ function DiscretizedDomain(
...
+        x3, x2, x1 = f(R, b / c, a / c)
+    elseif fdim == 3 && c < 0
+        # ( *, *,-R) : Face VI from Ronchi, Iacono, Paolucci (1996)
+        x3, x2, x1 = f(-R, b / c, a / c)
+    else
+        error("invalid case for cubedshellwarp: $a, $b, $c")
+    end
+
+    return x1, x2, x3
+end
```
Comment (charleskawczynski) [inline test/Numerics/DGMethods/compressible_navier_stokes_equations/shared_source/grids.jl:236]: This exact code snippet already exists exists [here](https://github.com/CliMA/ClimateMachine.jl/blob/master/test/Numerics/Mesh/filter.jl#L401-L434). Can we somehow share this instead?

Comment (sandreza) [reply test/Numerics/DGMethods/compressible_navier_stokes_equations/shared_source/grids.jl:236]: > This exact code snippet already exists exists [here](https://github.com/CliMA/ClimateMachine.jl/blob/master/test/Numerics/Mesh/filter.jl#L401-L434). Can we somehow share this instead?

We should definitely share it. The code snippet is already in https://github.com/CliMA/ClimateMachine.jl/blob/master/src/Numerics/Mesh/Topologies.jl#L2022. The keyword arguments made it so that the function could not be called in a way that repeated previous behavior. I'm open to suggestions on the best way of moving forward on that

File: test/Numerics/DGMethods/compressible_navier_stokes_equations/sphere/grid_notes.jl [hunk 3]
```diff
@@ -0,0 +1,38 @@
...
+# second index is ids
+# third index is elements
+# the are are 16 ids
+# 1-9 are metric terms
+# 10 is the mass matrix
+# 11 is the inverse mass matrix
+# 12 is MHid (horizontal mass matrix?)
+# 13-15 are x1 x2 x3
+# 16 is the vertical volume jacobian
+# grid.vgeo[:, grid.ξ1x1id, :]
```
Comment (blallen) [inline test/Numerics/DGMethods/compressible_navier_stokes_equations/sphere/grid_notes.jl:38]: @sandreza @charleskawczynski we should probably add this to official documentation somewhere

Comment (jkozdon) [reply test/Numerics/DGMethods/compressible_navier_stokes_equations/sphere/grid_notes.jl]: @valeriabarra is reworking all this as well in #2162

**Comment by blallen** [comment]
bors try

File: src/Numerics/Mesh/Topologies.jl [hunk 4]
```diff
@@ -2024,27 +2024,12 @@ function cubedshelltopowarp(
     b,
     c,
     R = max(abs(a), abs(b), abs(c));
-    r_inner = _planet_radius,
-    r_outer = _planet_radius + domain_height,
+    r_inner = 0,
+    r_outer = 1,
     topography::AnalyticalTopography = NoTopography(),
 )
+    f = get_topography_function(topography, r_inner, r_outer)
```
Comment (blallen) [inline src/Numerics/Mesh/Topologies.jl:2031]: @charleskawczynski here's how I handled the code duplication

File: test/Numerics/DGMethods/compressible_navier_stokes_equations/three_dimensional/ThreeDimensionalCompressibleNavierStokesEquations.jl (hunk 1 above)
Comment (blallen) [reply test/Numerics/DGMethods/compressible_navier_stokes_equations/three_dimensional/ThreeDimensionalCompressibleNavierStokesEquations.jl:283]: @jkozdon any thoughts?

File: src/Numerics/Mesh/Topologies.jl (hunk 4 above)
Comment (charleskawczynski) [reply src/Numerics/Mesh/Topologies.jl:2031]: Why couldn't we call the existing function? Wasn't it exactly the same?

File: test/Numerics/DGMethods/compressible_navier_stokes_equations/shared_source/grids.jl (hunk 2 above)
Comment (valeriabarra) [reply test/Numerics/DGMethods/compressible_navier_stokes_equations/shared_source/grids.jl]: @sandreza , @blallen In PR #2060 I had renamed `cubedshellwarp` as  `cubed_sphere_warp` and so on for the unwarp version and added two specialized case wrappers for the multiple dispatch (I preferred using `snake_case` because it is more readable/accessible). I reckon that I didn't update the `cubedshelltopowarp` function because it was not clear to us in the atmos group how/where that was used. It would be nice if now that you are touching this, the interface and function names could be unified a bit. Thanks

**Review by bischtob — APPROVED** [review]

File: src/Numerics/Mesh/Topologies.jl (hunk 4 above)
Comment (blallen) [reply src/Numerics/Mesh/Topologies.jl:2031]: it relied on planet_radius and domain_height and explicitly called analytical topography... which is unnecessary if you don't have topography. also causes the code to crash if planet_radius and domain_height are undefined, which is pretty common in the numerics tests. 

but also see this comment by @valeriabarra: https://github.com/CliMA/ClimateMachine.jl/pull/2171#discussion_r611946215

File: test/Numerics/DGMethods/compressible_navier_stokes_equations/shared_source/grids.jl (hunk 2 above)
Comment (blallen) [reply test/Numerics/DGMethods/compressible_navier_stokes_equations/shared_source/grids.jl]: oh! is that where it went? in that case I can probably remove this commit https://github.com/CliMA/ClimateMachine.jl/pull/2171/commits/6d4c7eaba50683d11eabfda81cd9057d4536f3b4 and just call `cubed_sphere_warp`

**Comment by blallen** [comment]
bors try

File: src/Numerics/Mesh/Topologies.jl (hunk 4 above)
Comment (blallen) [reply src/Numerics/Mesh/Topologies.jl]: dropped this change based on @valeriabarra's comment

File: test/Numerics/DGMethods/compressible_navier_stokes_equations/shared_source/grids.jl (hunk 2 above)
Comment (blallen) [reply test/Numerics/DGMethods/compressible_navier_stokes_equations/shared_source/grids.jl]: it worked! yay fewer changes!

**Comment by blallen** [comment]
bors try

File: test/Numerics/DGMethods/compressible_navier_stokes_equations/shared_source/grids.jl (hunk 2 above)
Comment (valeriabarra) [reply test/Numerics/DGMethods/compressible_navier_stokes_equations/shared_source/grids.jl]: Thanks @blallen . So, the `cubedshelltopowarp` and `cubedshelltopounwarp` are not needed at all then? Should they be renamed at least to follow the current convention of the other ones?

Comment (valeriabarra) [reply test/Numerics/DGMethods/compressible_navier_stokes_equations/shared_source/grids.jl]: @blallen not sure if you saw this. I can rename them in a separate quick PR

Comment (valeriabarra) [reply test/Numerics/DGMethods/compressible_navier_stokes_equations/shared_source/grids.jl]: Added #2181

Comment (blallen) [reply test/Numerics/DGMethods/compressible_navier_stokes_equations/shared_source/grids.jl]: sorry for the delay; gmail has decided everything from github is spam :'(

I think the `topowarp` versions are different functions? so #2181 is probably the way to go
